### ISSUE TITLE
"dart foo.dart" doesn't run `pub get` if necessary

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#49090]
When you do `flutter run`, if the `pubspec.lock` is missing or out of date relative to the `pubspec.yaml`, it runs `pub get` implicitly. We should do that for Dart too, so that we don't get unexpected error messages about broken imports. (Or if for some reason we don't want to do that, we should at a minimum warn that the dependencies are out of date and how to fix it, before printing error messages. But I think just doing it implicitly is probably the better experience.)

**Comment by mit-mit** [comment]
Yeah, looks like neither `dart foo.dart` not `dart run foo.dart` does this. cc @jonasfj @bkonyi I vaguely recall some discussion of this -- is there a reason why we don't do that?

**Comment by bkonyi** [comment]
There's a few reasons why we wouldn't want to do this that I can think of:

- `dart foo.dart` doesn't take the same path as `dart run foo.dart`. If the VM service isn't enabled, `dart foo.dart` short circuits and bypasses the CLI isolate completely to improve startup latency. I don't think we want to lose this behavior.
- Running `pub get` can (does?) hit the network, which can cause further startup latency issues if network connection is poor.

IMO, users have different expectations for the Flutter and Dart CLIs. Users of the Dart CLI expect for it to start as quickly as possible for production use (we've had complaints about startup latency in the past), whereas the Flutter CLI is solely used for development purposes where initial startup time may not be as important.

**Comment by jonasfj** [comment]
It's a bit inconsistent, because if you use the `dart run [package:]<command>` then `dart run` will do an implicit `dart pub get` (if needed).

This done implicitly by [`getExecutableForCommand`](https://github.com/dart-lang/pub/blob/master/lib/pub.dart#L10).

IMO, we should this logic for `dart run`, and expose a `getDependenciesIfNeeded` from:
https://github.com/dart-lang/pub/blob/master/lib/pub.dart

In the `pub` package we have logic for doing this really fast.

In practice, I see no use to trying to run `foo.dart` when dependencies are incorrect, because:
 * It yields non-sensible error messages,
 * The [cost of checking if dependencies are up-to-date](https://github.com/dart-lang/pub/blob/0ad17e8424e9989722f272e012636d9f43bb5205/lib/src/entrypoint.dart#L517) is fairly low:
     * We check timestamps,
     * We check if `pubspec_overrides.yaml` exists,
     * We check SDK constraints in `pubspec.lock` allows the current SDK (this is done with a cheap regular expression),
     * We read `pubspec.lock` as string and grep for path-dependencies, if there are path dependencies there is a slightly more expensive check, but it doesn't involve doing network.
  * If the dependencies are not satisfied, we can choose between:
    * Option A: fetch dependencies for the user.
    * Option B: exit with error and telling the user to fetch dependencies.

With this happening in the `dartdev` isolate where `pub` is now embedded, I think it's very cheap to do these things.
Historically, `pub run` (and by implication `pub run test`) used to do [`assertUpToDate`](https://github.com/dart-lang/pub/blob/0ad17e8424e9989722f272e012636d9f43bb5205/lib/src/entrypoint.dart#L517).


Don't we also want `dart analyze` to implicitly fetch dependencies, if they are not satisfied? That would stop a lot of messy errors.

**Comment by Hixie** [comment]
The only time we'd need to hit the network and take a latency hit are cases where the dependencies are out of date (pubspec.yaml updated more recently than pubspec.lock, or pubspec.lock doesn't exist at all), and in those cases, running the code is semantically wrong, IMHO. Half the time you'll just get error messages anyway (in which case who cares about the latency hit) and the other half of the time the code you're running is going to be a weird non-reproducible state that is just as likely to be buggy as correct and if it's buggy we're just going to send people on wild goose chases if we run it.

Certainly as a user I would _much_ rather take the latency hit on the first run after updating `pubspec.yaml` or checking out the code than spending minutes trying to debug weird error messages or buggy behaviour.

FWIW I think to the extent that `dart run foo` and `dart foo` do different things we're just confusing developers. I certainly can't explain what the difference is...
